File: SSMConnectKit/Tests/SSMConnectKitTests/Models/PortableProfileTests.swift
import Foundation
import Testing
@testable import SSMConnectKit

/// Covers the portable profile document (spec §6.1, AC-03).
///
/// The important tests here are the ones that read `contracts/fixtures/profiles/*.json` rather than
/// a locally-authored string. Those files are the shared contract and the .NET client's fixture
/// loader reads the same three documents, so a change on either side that breaks interchange fails
/// here instead of at the first real cross-platform import.
@Suite("Portable profile document")
struct PortableProfileTests {
    static var profilesDirectory: URL {
        FixturePaths.contractsDirectory.appendingPathComponent("fixtures/profiles")
    }

    static func fixtureData(_ name: String) throws -> Data {
        try Data(contentsOf: profilesDirectory.appendingPathComponent("\(name).json"))
    }

    static let fixtureNames = ["single-user", "multi-user", "unconfigured"]

    // MARK: Importing the real contract fixtures

    @Test("every contract profile fixture imports", arguments: fixtureNames)
    func contractFixturesImport(name: String) throws {
        let document = try ProfilePortability.importDocument(from: Self.fixtureData(name))
        #expect(document.schemaVersion == 1)
        #expect(!document.name.isEmpty)
    }

    @Test("the single-user fixture maps onto the expected profile")
    func singleUserMapping() throws {
        let profile = try ProfilePortability.importProfile(from: Self.fixtureData("single-user"))

        #expect(profile.id == UUID(uuidString: "11111111-1111-4111-8111-111111111111"))
        #expect(profile.name == "Example Single-User Workstation")
        #expect(profile.ssoRegion == "eu-west-1")
        #expect(profile.accountId == "000000000000")
        #expect(profile.resourceRegion == "eu-central-1")
        #expect(profile.instanceTagValue == "example-workstation-su")
        #expect(profile.secretId == "example/dcv/password")
        #expect(profile.localPort == 8443)
        #expect(profile.remotePort == 8443)
        #expect(profile.resolvedConnectMode == .singleUser)
        // Not a multi-user profile, so the agent port stays at its default rather than being stored.
        #expect(profile.agentRemotePort == nil)
        #expect(profile.resolvedAgentRemotePort == 8444)
    }

    @Test("the multi-user fixture maps onto the expected profile")
    func multiUserMapping() throws {
        let profile = try ProfilePortability.importProfile(from: Self.fixtureData("multi-user"))

        #expect(profile.resolvedConnectMode == .multiUser)
        #expect(profile.agentRemotePort == 8444)
        // MU-00a: a multi-user host is identity-only. The fixture carries an explicit null.
        #expect(profile.secretId == nil)
        #expect(profile.instanceTagValue == "example-workstation-mu")
    }

    @Test("the unconfigured fixture imports and is correctly reported as not connectable")
    func unconfiguredMapping() throws {
        let profile = try ProfilePortability.importProfile(from: Self.fixtureData("unconfigured"))

        // An empty instanceTagValue is schema-valid: a profile can be exported before it is finished.
        #expect(profile.instanceTagValue.isEmpty)
        #expect(!profile.isConfigured)
    }

    // MARK: Round-trip

    @Test("import then export reproduces the contract fixture exactly", arguments: fixtureNames)
    func roundTripPreservesTheDocument(name: String) throws {
        let original = try Self.fixtureData(name)
        let document = try ProfilePortability.importDocument(from: original)
        let exported = try ProfilePortability.export(document.profile, unknownFields: document.unknownFields)

        let before = try JSONSerialization.jsonObject(with: original) as? [String: Any] ?? [:]
        let after = try JSONSerialization.jsonObject(with: exported) as? [String: Any] ?? [:]

        // Compared as parsed JSON, not as text: key order and whitespace are not contract.
        #expect(
            NSDictionary(dictionary: normalize(before)) == NSDictionary(dictionary: normalize(after)),
            "Round-tripping '\(name)' changed the document.\n  before: \(normalize(before))\n  after:  \(normalize(after))"
        )
    }

    /// An explicit `"secretId": null` and an absent `secretId` mean the same thing to both clients,
    /// so the comparison treats them as equal rather than forcing the exporter to write nulls.
    private func normalize(_ object: [String: Any]) -> [String: Any] {
        object.filter { !($0.value is NSNull) }
    }

    @Test("a profile with no connectMode round-trips without gaining one")
    func absentConnectModeStaysAbsent() throws {
        // This is the shape the shipping macOS app actually has stored: connectMode predates the
        // field, so a real export omits it. Resolving it to "singleUser" on export would silently
        // rewrite the document and make AC-03 pass against synthetic data only.
        var profile = ConnectionProfile.template
        profile.name = "Legacy"
        profile.ssoStartUrl = "https://d-0000000000.awsapps.com/start"
        profile.ssoRegion = "eu-west-1"
        profile.accountId = "000000000000"
        profile.roleName = "ExampleRole"
        profile.resourceRegion = "eu-central-1"
        profile.instanceTagValue = "example"
        profile.connectMode = nil

        let exported = try ProfilePortability.exportString(profile)
        #expect(!exported.contains("connectMode"))

        let reimported = try ProfilePortability.importProfile(from: Data(exported.utf8))
        #expect(reimported.connectMode == nil)
        #expect(reimported.resolvedConnectMode == .singleUser)
    }

    @Test("unknown optional properties survive a round-trip")
    func unknownFieldsArePreserved() throws {
        var object = try JSONSerialization.jsonObject(with: Self.fixtureData("single-user")) as! [String: Any]
        object["futureFieldFromANewerClient"] = "keep me"
        object["futureNumber"] = 7
        let data = try JSONSerialization.data(withJSONObject: object)

        let document = try ProfilePortability.importDocument(from: data)
        let exported = try ProfilePortability.export(document.profile, unknownFields: document.unknownFields)
        let result = try JSONSerialization.jsonObject(with: exported) as! [String: Any]

        #expect(result["futureFieldFromANewerClient"] as? String == "keep me")
        #expect((result["futureNumber"] as? NSNumber)?.intValue == 7)
    }

    // MARK: Export shape

    @Test("connectAction exports as a plain string, not a synthesized enum object")
    func connectActionEncodesAsAString() throws {
        // ConnectAction has no raw value, so ConnectionProfile's synthesized Codable would emit
        // {"dcvViewer":{}} here. That is schema-invalid, and it is the whole reason this document
        // is a separate type rather than a Codable conformance on the native model.
        let exported = try ProfilePortability.exportString(.template)
        #expect(exported.contains("\"connectAction\" : \"dcvViewer\""))
    }

    @Test("ports export as integers rather than floating-point numbers")
    func portsExportAsIntegers() throws {
        let exported = try ProfilePortability.exportString(.template)
        #expect(exported.contains("\"localPort\" : 8443"))
        #expect(!exported.contains("8443.0"))
    }

    @Test("an exported document always declares the current schema version")
    func exportDeclaresSchemaVersion() throws {
        let exported = try ProfilePortability.exportString(.template)
        #expect(exported.contains("\"schemaVersion\" : 1"))
    }

    // MARK: Rejections

    @Test("a newer schema version is refused with an actionable message")
    func futureSchemaVersionIsRefused() throws {
        var object = try JSONSerialization.jsonObject(with: Self.fixtureData("single-user")) as! [String: Any]
        object["schemaVersion"] = 2
        let data = try JSONSerialization.data(withJSONObject: object)

        #expect(throws: ProfilePortabilityError.unsupportedSchemaVersion(found: 2)) {
            try ProfilePortability.importDocument(from: data)
        }
    }

    @Test(
        "a field that violates the schema is refused",
        arguments: [
            ("accountId", "12345678901" as Any),      // 11 digits
            ("accountId", "00000000000a" as Any),     // right length, not all digits
            ("resourceRegion", "eu-central-1-" as Any),
            ("ssoRegion", "" as Any),
            ("localPort", 0 as Any),
            ("remotePort", 70000 as Any),
            ("name", "" as Any),
            ("connectMode", "someOtherMode" as Any),
            ("connectAction", "rdp" as Any),
            ("id", "not-a-uuid" as Any),
        ]
    )
    func invalidFieldsAreRefused(field: String, value: Any) throws {
        var object = try JSONSerialization.jsonObject(with: Self.fixtureData("single-user")) as! [String: Any]
        object[field] = value
        let data = try JSONSerialization.data(withJSONObject: object)

        #expect(throws: ProfilePortabilityError.self, "'\(field)' = \(value) should have been refused") {
            try ProfilePortability.importDocument(from: data)
        }
    }

    @Test("a required field that is missing is named in the error")
    func missingFieldIsNamed() throws {
        var object = try JSONSerialization.jsonObject(with: Self.fixtureData("single-user")) as! [String: Any]
        object.removeValue(forKey: "roleName")
        let data = try JSONSerialization.data(withJSONObject: object)

        #expect(throws: ProfilePortabilityError.missingField("roleName")) {
            try ProfilePortability.importDocument(from: data)
        }
    }

    @Test("a multi-user profile carrying a secret ID is refused")
    func multiUserWithSecretIsRefused() throws {
        var object = try JSONSerialization.jsonObject(with: Self.fixtureData("multi-user")) as! [String: Any]
        object["secretId"] = "example/dcv/password"
        let data = try JSONSerialization.data(withJSONObject: object)

        #expect(throws: ProfilePortabilityError.self) {
            try ProfilePortability.importDocument(from: data)
        }
    }

    @Test("a document carrying credential material is refused outright", arguments: [
        "accessKeyId", "secretAccessKey", "sessionToken", "password", "authToken", "ssoAccessToken",
    ])
    func credentialBearingDocumentIsRefused(key: String) throws {
        var object = try JSONSerialization.jsonObject(with: Self.fixtureData("single-user")) as! [String: Any]
        object[key] = "should-never-be-imported"
        let data = try JSONSerialization.data(withJSONObject: object)

        // Without this, the unknown-field preservation above would happily carry a secret through.
        #expect(throws: ProfilePortabilityError.forbiddenField(key)) {
            try ProfilePortability.importDocument(from: data)
        }
    }

    @Test("malformed input is refused rather than crashing")
    func malformedInputIsRefused() {
        #expect(throws: ProfilePortabilityError.malformedJSON) {
            try ProfilePortability.importDocument(from: Data("{ not json".utf8))
        }
        #expect(throws: ProfilePortabilityError.notAnObject) {
            try ProfilePortability.importDocument(from: Data("[]".utf8))
        }
    }

    // MARK: The macOS half of AC-03

    @Test("a profile exported from the app re-imports to an identical profile")
    func exportedProfileReimportsIdentically() throws {
        for name in Self.fixtureNames {
            let original = try ProfilePortability.importProfile(from: Self.fixtureData(name))
            let exported = try ProfilePortability.export(original)
            let reimported = try ProfilePortability.importProfile(from: exported)

            #expect(reimported == original, "'\(name)' did not survive export → import unchanged.")
        }
    }
}

File: windows/tests/SSMConnect.Domain.Tests/PortableProfileTests.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using SSMConnect.Domain;

namespace SSMConnect.Domain.Tests;

/// <summary>
/// Covers the portable profile document (specification §6.1). The round-trip cases run against the
/// real contract fixtures rather than hand-written JSON, and every exported document is written to
/// <c>artifacts/exported-profiles</c> so <c>contracts/validate.py</c> can check it against the
/// schema — a .NET assertion cannot catch a schema violation such as an enum written as a number.
/// </summary>
public sealed class PortableProfileTests
{
    private static string FixturesDirectory =>
        Path.Combine(AppContext.BaseDirectory, "contracts", "fixtures", "profiles");

    private static string ArtifactsDirectory
    {
        get
        {
            string path = Path.Combine(AppContext.BaseDirectory, "artifacts", "exported-profiles");
            Directory.CreateDirectory(path);
            return path;
        }
    }

    public static TheoryData<string> ProfileFixtures()
    {
        var data = new TheoryData<string>();
        foreach (string path in Directory.GetFiles(FixturesDirectory, "*.json").OrderBy(p => p, StringComparer.Ordinal))
        {
            data.Add(Path.GetFileName(path));
        }

        return data;
    }

    [Fact]
    public void TheContractProfileFixturesAreAvailable() =>
        Assert.True(Directory.GetFiles(FixturesDirectory, "*.json").Length >= 3,
            $"Expected the contract profile fixtures next to the tests, in '{FixturesDirectory}'.");

    /// <summary>
    /// Import then export then import must reach the same document. This is the property that makes
    /// a profile safe to move between clients repeatedly rather than only once.
    /// </summary>
    [Theory]
    [MemberData(nameof(ProfileFixtures))]
    public void RoundTripsAContractFixture(string fileName)
    {
        string original = File.ReadAllText(Path.Combine(FixturesDirectory, fileName));

        PortableProfileDocument first = ProfilePortability.Import(original);
        string exported = ProfilePortability.Export(first);
        PortableProfileDocument second = ProfilePortability.Import(exported);

        Assert.Equal(first.Id, second.Id);
        Assert.Equal(first.Name, second.Name);
        Assert.Equal(first.SsoStartUrl, second.SsoStartUrl);
        Assert.Equal(first.SsoRegion, second.SsoRegion);
        Assert.Equal(first.AccountId, second.AccountId);
        Assert.Equal(first.RoleName, second.RoleName);
        Assert.Equal(first.ResourceRegion, second.ResourceRegion);
        Assert.Equal(first.InstanceTagKey, second.InstanceTagKey);
        Assert.Equal(first.InstanceTagValue, second.InstanceTagValue);
        Assert.Equal(first.LocalPort, second.LocalPort);
        Assert.Equal(first.RemotePort, second.RemotePort);
        Assert.Equal(first.ConnectMode, second.ConnectMode);
        Assert.Equal(first.AgentRemotePort, second.AgentRemotePort);
        Assert.Equal(first.SecretId, second.SecretId);
        Assert.Equal(first.ConnectAction, second.ConnectAction);

        // Hand the exported document to validate.py, which owns the schema.
        File.WriteAllText(Path.Combine(ArtifactsDirectory, fileName), exported);
    }

    /// <summary>
    /// The whole reason this is a separate type: the enum must reach the file as a string. Left to
    /// System.Text.Json's defaults on the native model it would be written as a number.
    /// </summary>
    [Fact]
    public void WritesEnumsAsSchemaStringsNotNumbers()
    {
        string json = ProfilePortability.Export(Example() with { ConnectMode = ConnectMode.MultiUser, SecretId = null });
        JsonNode root = JsonNode.Parse(json)!;

        Assert.Equal(JsonValueKind.String, root["connectAction"]!.GetValueKind());
        Assert.Equal("dcvViewer", root["connectAction"]!.GetValue<string>());
        Assert.Equal(JsonValueKind.String, root["connectMode"]!.GetValueKind());
        Assert.Equal("multiUser", root["connectMode"]!.GetValue<string>());
    }

    /// <summary>
    /// A profile stored before the connect-mode field existed has no such key. Writing a resolved
    /// value would rewrite the user's document, so absent must stay absent.
    /// </summary>
    [Fact]
    public void AnAbsentConnectModeStaysAbsentOnReExport()
    {
        string json = ProfilePortability.Export(Example());
        JsonObject root = (JsonObject)JsonNode.Parse(json)!;

        Assert.False(root.ContainsKey("connectMode"));
        Assert.False(root.ContainsKey("agentRemotePort"));
        Assert.Null(ProfilePortability.Import(json).ConnectMode);

        // Resolution is still available to callers; it just is not baked into the document.
        Assert.Equal(ConnectMode.SingleUser, ProfilePortability.ImportProfile(json).ResolvedConnectMode);
        Assert.Equal(8444, ProfilePortability.ImportProfile(json).ResolvedAgentRemotePort);
    }

    [Fact]
    public void PreservesUnknownOptionalProperties()
    {
        string source = ProfilePortability.Export(Example());
        JsonObject root = (JsonObject)JsonNode.Parse(source)!;
        root["futureFeature"] = "some-value";
        root["futureCount"] = 7;

        PortableProfileDocument imported = ProfilePortability.Import(root.ToJsonString());
        Assert.Equal(2, imported.UnknownFields.Count);

        JsonObject reExported = (JsonObject)JsonNode.Parse(ProfilePortability.Export(imported))!;
        Assert.Equal("some-value", reExported["futureFeature"]!.GetValue<string>());
        Assert.Equal(7, reExported["futureCount"]!.GetValue<int>());
    }

    /// <summary>
    /// Preservation stops at the schema's security clause. A document carrying credential material
    /// must be rejected, not carried through as an unknown optional property.
    /// </summary>
    [Theory]
    [InlineData("accessKeyId")]
    [InlineData("secretAccessKey")]
    [InlineData("sessionToken")]
    [InlineData("password")]
    [InlineData("authToken")]
    [InlineData("presignedUrl")]
    [InlineData("connectionFilePath")]
    public void RejectsADocumentCarryingCredentialMaterial(string forbiddenKey)
    {
        JsonObject root = (JsonObject)JsonNode.Parse(ProfilePortability.Export(Example()))!;
        root[forbiddenKey] = "should-never-be-imported";

        var error = Assert.Throws<ProfilePortabilityException>(
            () => ProfilePortability.Import(root.ToJsonString()));
        Assert.Contains(forbiddenKey, error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void EveryForbiddenKeyInTheSchemaIsEnforced()
    {
        // Kept in step with the schema's deny-list; a key added there must be added here.
        string[] expected =
        [
            "accessKeyId", "secretAccessKey", "sessionToken", "credentials",
            "accessToken", "refreshToken", "ssoAccessToken",
            "password", "dcvPassword", "secretValue", "authToken", "presignedUrl",
            "sessionResponse", "ssmSession", "connectionFilePath", "tempFilePath",
        ];

        Assert.Equal(expected.Order(StringComparer.Ordinal), ProfilePortability.ForbiddenKeys.Order(StringComparer.Ordinal));
    }

    [Fact]
    public void RejectsAnUnsupportedSchemaVersionWithAnActionableMessage()
    {
        JsonObject root = (JsonObject)JsonNode.Parse(ProfilePortability.Export(Example()))!;
        root["schemaVersion"] = 2;

        var error = Assert.Throws<ProfilePortabilityException>(
            () => ProfilePortability.Import(root.ToJsonString()));
        Assert.Contains("version 2", error.Message, StringComparison.Ordinal);
        Assert.Contains("Update SSM Connect", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void RejectsAMultiUserProfileCarryingASecretId()
    {
        JsonObject root = (JsonObject)JsonNode.Parse(
            ProfilePortability.Export(Example() with { ConnectMode = ConnectMode.MultiUser }))!;
        root["secretId"] = "example/dcv/password";

        var error = Assert.Throws<ProfilePortabilityException>(
            () => ProfilePortability.Import(root.ToJsonString()));
        Assert.Contains("no shared password", error.Message, StringComparison.Ordinal);
    }

    [Theory]
    [InlineData("accountId", "\"12345\"", "12 digits")]
    [InlineData("resourceRegion", "\"eu-central-1-\"", "valid AWS region")]
    [InlineData("localPort", "0", "between 1 and 65535")]
    [InlineData("name", "\"\"", "must not be empty")]
    public void RejectsInvalidFieldsNamingTheFieldAndValue(string field, string rawValue, string reasonFragment)
    {
        JsonObject root = (JsonObject)JsonNode.Parse(ProfilePortability.Export(Example()))!;
        root[field] = JsonNode.Parse(rawValue);

        var error = Assert.Throws<ProfilePortabilityException>(
            () => ProfilePortability.Import(root.ToJsonString()));
        Assert.Contains(field, error.Message, StringComparison.Ordinal);
        Assert.Contains(reasonFragment, error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void RejectsAMissingRequiredField()
    {
        JsonObject root = (JsonObject)JsonNode.Parse(ProfilePortability.Export(Example()))!;
        root.Remove("roleName");

        var error = Assert.Throws<ProfilePortabilityException>(
            () => ProfilePortability.Import(root.ToJsonString()));
        Assert.Contains("roleName", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void RejectsMalformedJson() =>
        Assert.Throws<ProfilePortabilityException>(() => ProfilePortability.Import("{not json"));

    [Fact]
    public void RejectsJsonThatIsNotAnObject() =>
        Assert.Throws<ProfilePortabilityException>(() => ProfilePortability.Import("[]"));

    /// <summary>An import failure is a configuration problem, so it classifies as one.</summary>
    [Fact]
    public void AnImportFailureClassifiesAsConfiguration() =>
        Assert.Equal(ErrorCategory.Configuration,
            ErrorCategories.Classify(ProfilePortabilityException.MalformedJson()));

    private static ConnectionProfile Example() => new()
    {
        Id = Guid.Parse("11111111-1111-4111-8111-111111111111"),
        Name = "Example Workstation",
        SsoStartUrl = "https://d-0000000000.awsapps.com/start",
        SsoRegion = "eu-west-1",
        AccountId = "000000000000",
        RoleName = "ExampleRole",
        ResourceRegion = "eu-central-1",
        InstanceTagKey = "Name",
        InstanceTagValue = "example-workstation",
        LocalPort = 8443,
        RemotePort = 8443,
    };
}
